docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,341 @@
+let mut inst = region
+    .new_instance(module)
+    .expect("instance can be created");
+let kill_switch = inst.kill_switch();
+
+// Spawn a thread to terminate the instance after waiting for 100ms.
+let t = thread::Builder::new()
+    .name("killswitch".to_owned())
+    .spawn(move || {
+        thread::sleep(Duration::from_millis(100));
+        assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+    })
+    .expect("can spawn a thread");
+
+// Begin running the instance, which will be terminated remotely by the KillSwitch.
+match inst.run("infinite_loop", &[]) {
+    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
+        // the result of a guest that was remotely terminated (a KillSwitch at work)
+    }
+    res => panic!("unexpected result: {:?}", res),
+}
+
+// wait for the KillSwitch-firing thread to complete
+t.join().unwrap();
+```
+
+## Implementation
+
+As this section discusses implementation details of `lucet-runtime`, it
+will refer to structures and enums that are unexported.  Since there are
+no `lucet-runtime` paths for these items, this section will occasionally
+use fully-qualified paths to the `lucet_runtime_internals` path where.
+they are originally defined.
+
+As much as is possible, `KillSwitch` tries to be self-contained; no members are
+public, and it tries to avoid leaking details of its state-keeping into public
+interfaces. Currently, `lucet-runtime` is heavily dependent on POSIX
+thread-directed signals to implement guest timeouts - for non-POSIX platforms
+alternate implementations may be plausible, but need careful consideration of
+the race conditions that can arise from other platform-specific functionality.
+
+`KillSwitch` fundamentally relies on two piece of state for safe operation,
+which are encapsulated in a `KillState` held by the `lucet_runtime::Instance`
+it terminates:
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`. This is kept in a `Mutex` since in
+  many cases it will need to be accessed either by a `KillSwitch` or
+  `KillState`, and for the duration either are considering the domain, it must
+  block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
+
+The astute observer may notice that `Lucet` contains both an instance `Domain`
+and an instance `State`. These discuss different aspects of the instance's
+lifecycle: `Domain` describes what the instance is doing right now, where
+`State` describes the state of the instance as visible outside
+`lucet_runtime`. This is fundamentally the reason why `State` does not and
+cannot describe a "making a hostcall" variant - this is not something
+`lucet_runtime` wants to expose, and at the moment we don't intend to make any
+commitments about being able to query instance-internal state like this.
+Additionally, this is why `Domain` only expresses that an instance has
```

Comment:
I don't understand the logical jump being made here.  Why does Domain only express that an instance has stopped?

Reply:
Oops, this was poor phrasing. `Domain` only has the ability to express that an instance has stopped, not how. Specifically, `Domain` has `Domain::Terminated` but not `Domain::Faulted`.

I'll rephrase this to not suggest that `Domain` specifically expresses an instance has stopped, and instead emphasize that `Domain::Terminated` implies the instance has stopped, and that details for that termination are over on the instance's `State`.